#### 1. Загрузите датасет ирисы Фишера из библиотеки sklearn.datasets.

In [ ]:
from sklearn.datasets import load_iris

iris = load_iris()
X = iris.data
y = iris.target

X, y


(array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
        [5

#### 2. Сделайте hold-out разбиение данных. Для этого разделите данные на обучающую и валидационную выборки и выведите на экран соответствующие индексы разбиения.

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_validation, y_train, y_validation = train_test_split(X, y, test_size=0.2, random_state=42)

# Получение индексов обучающей и валидационной выборок
train_indexes = X_train.index if hasattr(X_train, 'index') else range(len(X_train))
validation_indexes = X_validation.index if hasattr(X_validation, 'index') else range(len(X_train), len(X_train) + len(X_validation))

train_indexes, validation_indexes


(range(0, 120), range(120, 150))

####3. Теперь сделайте разбиение перемешанных данных, зафиксировав воспроизводимость выбора данных после перемешивания, указав значение параметра random_state=42 и выведите на экран соответствующие индексы разбиения.

In [ ]:
X_train_shuffled, X_validation_shuffled, y_train_shuffled, y_validation_shuffled = train_test_split(X, y, test_size=0.2, random_state=42, shuffle=True)

# Поскольку индексы не доступны напрямую (данные являются numpy массивами), используем альтернативный способ определения индексов через mask
# Создадим полный массив индексов
full_indexes = range(len(X))

# Находим обучающие индексы как те, которые не в валидационном наборе, и наоборот
train_mask = np.isin(full_indexes, train_indexes, invert=True)
validation_mask = np.isin(full_indexes, validation_indexes)

# Получаем исходные индексы для перемешанных выборок
train_shuffled_indexes = np.array(full_indexes)[train_mask]
validation_shuffled_indexes = np.array(full_indexes)[validation_mask]

train_shuffled_indexes, validation_shuffled_indexes


(array([120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132,
        133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145,
        146, 147, 148, 149]),
 array([120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132,
        133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145,
        146, 147, 148, 149]))

####4. Обучите модель логистической регрессии на обучающих данных. Выведите значения коэффициентов модели, полученных в результате обучения. Сделайте предсказание на тестовом наборе признаков. Выведите значение метрик accuracy и f1-score.

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, f1_score

model = LogisticRegression(max_iter=200, solver='liblinear')
model.fit(X_train_shuffled, y_train_shuffled)

coefficients = model.coef_

y_pred = model.predict(X_validation_shuffled)

accuracy = accuracy_score(y_validation_shuffled, y_pred)
f1 = f1_score(y_validation_shuffled, y_pred, average='macro')

print("Коэффициенты модели логистической регрессии:\n")
classes = ["Iris-setosa", "Iris-versicolor", "Iris-virginica"]
for index, class_name in enumerate(classes):
    print(f"{class_name}: {coefficients[index]}")
print("\nМетрики на валидационном наборе данных:")
print(f"Accuracy: {accuracy:.4f}")
print(f"F1-Score (macro): {f1:.4f}")



Коэффициенты модели логистической регрессии:

Iris-setosa: [ 0.3711229   1.409712   -2.15210117 -0.95474179]
Iris-versicolor: [ 0.49400451 -1.58897112  0.43717015 -1.11187838]
Iris-virginica: [-1.55895271 -1.58893375  2.39874554  2.15556209]

Метрики на валидационном наборе данных:
Accuracy: 1.0000
F1-Score (macro): 1.0000


####5. Разделите данные на обучающую и валидационную выборки по новому в соотношении 75-25. Обучите модель на этих данных, выведите значения получившихся коэффициентов модели. Выведите значения метрик и сравните их со значениями из предыдущего пункта. Сделайте вывод о том, влияет ли способ разбиения на результат.

In [ ]:
X_train_new, X_validation_new, y_train_new, y_validation_new = train_test_split(X, y, test_size=0.25, random_state=42)

model_new = LogisticRegression(max_iter=200, solver='liblinear')
model_new.fit(X_train_new, y_train_new)

coefficients_new = model_new.coef_

y_pred_new = model_new.predict(X_validation_new)

accuracy_new = accuracy_score(y_validation_new, y_pred_new)
f1_new = f1_score(y_validation_new, y_pred_new, average='macro')

print("Коэффициенты модели логистической регрессии после нового разбиения данных (75-25):\n")
for index, class_name in enumerate(classes):
    print(f"{class_name}: {coefficients_new[index]}")
print("\nМетрики на новом валидационном наборе данных:")
print(f"Accuracy: {accuracy_new:.4f}")
print(f"F1-Score (macro): {f1_new:.4f}")



Коэффициенты модели логистической регрессии после нового разбиения данных (75-25):

Iris-setosa: [ 0.37199402  1.3721129  -2.12209543 -0.93577108]
Iris-versicolor: [ 0.46579126 -1.55972584  0.41868466 -1.08288284]
Iris-virginica: [-1.55311025 -1.5151226   2.36636048  2.1109172 ]

Метрики на новом валидационном наборе данных:
Accuracy: 1.0000
F1-Score (macro): 1.0000


Изменение соотношения разбиения данных на обучающую и валидационную выборки не оказало заметного влияния на результаты классификации модели логистической регрессии в данном случае.

####6. Теперь сделайте k-блочную перекрёстную проверку модели (кросс-валидацию). Сравните полученные метрики с метриками, которые были при hold-out разбиении.

In [ ]:
from sklearn.model_selection import cross_val_score

model_cv = LogisticRegression(max_iter=200, solver='liblinear')

# Выполнение k-блочной перекрёстной проверки для метрики accuracy
cv_accuracy = cross_val_score(model_cv, X, y, cv=5, scoring='accuracy')

# Выполнение k-блочной перекрёстной проверки для метрики f1-score с усреднением по макро
cv_f1 = cross_val_score(model_cv, X, y, cv=5, scoring='f1_macro')

cv_accuracy_mean = np.mean(cv_accuracy)
cv_f1_mean = np.mean(cv_f1)

cv_accuracy_mean, cv_f1_mean


(0.9600000000000002, 0.959522933505973)

При предыдущем hold-out разбиении данных мы получили идеальные значения метрик (Accuracy и F1-Score) равные 1.0. В случае кросс-валидации, значения метрик немного ниже, но всё же очень высоки.

Кросс-валидация предоставляет более обобщенную оценку качества модели, так как данные оцениваются на нескольких разбиениях, а результаты усредняются. Это уменьшает влияние конкретного разбиения данных на оценку модели и предоставляет более надежную оценку её способности к обобщению на новых данных.

####7. Теперь сделайте ту же самую перекрёстную проверку модели, используя библиотечную функцию cross_val_score. Убедитесь, что получится тот же результат.

In [ ]:
cv_accuracy_repeat = cross_val_score(model_cv, X, y, cv=5, scoring='accuracy')

cv_f1_repeat = cross_val_score(model_cv, X, y, cv=5, scoring='f1_macro')

cv_accuracy_mean_repeat = np.mean(cv_accuracy_repeat)
cv_f1_mean_repeat = np.mean(cv_f1_repeat)

cv_accuracy_mean_repeat, cv_f1_mean_repeat


(0.9600000000000002, 0.959522933505973)

Это подтверждает надёжность и воспроизводимость результатов кросс-валидации с использованием библиотечной функции. Полученные метрики согласуются с предыдущими значениями, что демонстрирует высокое качество модели и её способность к обобщению на новых данных.

####8. Теперь сделайте k-блочную перекрёстную проверку модели (кросс-валидацию) со стратификацией. Проделайте всё тоже самое, что и в предыдущем пункте.

In [ ]:
from sklearn.model_selection import StratifiedKFold

cv_stratified = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

cv_accuracy_stratified = cross_val_score(model_cv, X, y, cv=cv_stratified, scoring='accuracy')

cv_f1_stratified = cross_val_score(model_cv, X, y, cv=cv_stratified, scoring='f1_macro')

cv_accuracy_stratified_mean = np.mean(cv_accuracy_stratified)
cv_f1_stratified_mean = np.mean(cv_f1_stratified)

cv_accuracy_stratified_mean, cv_f1_stratified_mean


(0.96, 0.959522933505973)

Значения метрик остались такими же, как и при обычной кросс-валидации без стратификации, что указывает на стабильность модели и её способность к обобщению на новых данных, независимо от метода разбиения.

####9. Теперь сделайте перекрёстную проверку, изпользуя leave-one-out разбиение. Проделайте всё тоже самое, что и в предыдущем пункте.

In [ ]:
from sklearn.model_selection import LeaveOneOut

# Создание разбиения leave-one-out
cv_loo = LeaveOneOut()

cv_accuracy_loo = cross_val_score(model_cv, X, y, cv=cv_loo, scoring='accuracy')

cv_f1_loo = cross_val_score(model_cv, X, y, cv=cv_loo, scoring='f1_macro')

cv_accuracy_loo_mean = np.mean(cv_accuracy_loo)
cv_f1_loo_mean = np.mean(cv_f1_loo)

cv_accuracy_loo_mean, cv_f1_loo_mean


(0.9533333333333334, 0.9533333333333334)

Значения метрик немного ниже, чем при использовании стратифицированной и нестратифицированной k-блочной перекрёстной проверки (где оба показателя были равны 0.96).

Метод leave-one-out предоставляет один из самых строгих способов кросс-валидации, поскольку модель обучается на всех данных, кроме одного элемента, а затем тестируется на этом единственном элементе. Это процесс повторяется для каждого элемента в наборе данных. Такой подход может быть более чувствителен к выбросам и шуму в данных, что может объяснить незначительное снижение метрик по сравнению с другими методами кросс-валидации. Несмотря на это, полученные результаты всё ещё указывают на высокую эффективность модели. ​